Use Application Default Credentials to get App/Compute Engine credentials

ajkannan opened the pull request:
Fixes #334 

2 notes:
-- This PR should be merged in only once #336 is merged in
-- I left the order for project ID determination the same for now.  There was a suggestion to make it the same order as the authentication credentials lookup.  However, since we check for App Engine ID before Google Cloud SDK ID, we don't run into security exceptions with using the files API if we run in App Engine.  However, if we change the order to move Google Cloud SDK first, we have to try and catch that exception, and the effective order would be slightly different than what we might report.  Somebody might change their JRE locally to allow read access to the Google Cloud SDK config file, but then it would crash in production.

aozarov commented:
I am not sure I understand your comment about the order.
What we want is to switch the order of project default to look up at "Google Cloud SDK"
before "Compute Engine" which is similar to the lookup order for auth. In both cases (auth or projectId) AE is picked before and therefore we should not have the file access issue.

aozarov commented:
Ah, I get you now. This PR actually changed the lookup order of auth as we now rely on google-auth-library-java for AE auth. I think we could still swith the projectid lookup order of CE with SDK, no?

aozarov commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/SerializationTest.java:
```diff
@@ -133,20 +133,21 @@
-        .authCredentials(AuthCredentials.noCredentials())
-        .force(true)
-        .build();
+    options =
```
is that wrapping preferred or was it necessary (style, formatter)?

ajkannan replied:
They were preferred, not necessary.  I reverted that style change here and in storage's SerializationTest.

ajkannan commented:
@aozarov Added a commit to switch the project ID lookup order of CE and SDK.

ajkannan commented:
Waiting on this [bug](https://github.com/google/google-auth-library-java/issues/43) to be fixed, will test and update PR as necessary following that fix.

ajkannan commented:
Yup, that's correct.

ajkannan commented:
This PR now is only blocked by [this feature request](https://github.com/google/google-auth-library-java/issues/50) from the auth library.

aozarov commented:
Once we resolve this issue we should also fix  #334

ajkannan commented:
Added back in the app engine credentials since it looks like reflection won't be added to the auth library.

aozarov commented on gcloud-java-core/src/main/java/com/google/gcloud/ServiceOptions.java:
```diff
@@ -111,12 +115,22 @@ public HttpTransport create() {
       // Consider Compute
       try {
-        return AuthCredentials.getComputeCredential().getTransport();
+        return getComputeHttpTransport();
```
I have the feeling that this is not necessary (but this needs to be validated).
Basically, I think `NetHttpTransport` works on GCE (consider [this](https://github.com/google/google-auth-library-java/blob/281d485e679545ed06f46e6090fc5d47f4a140e8/oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java#L45) and [this](https://github.com/google/google-api-java-client/blob/3d866eb7094cfd31ad9f7acefbadd6cd5171be48/google-api-client/src/main/java/com/google/api/client/googleapis/javanet/GoogleNetHttpTransport.java#L55)).

And the the when choosing the default auth (if not specified) it will be checked (see [this](https://github.com/google/google-auth-library-java/blob/281d485e679545ed06f46e6090fc5d47f4a140e8/oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java#L98).

If we need a way to check "isOnGCE" we can apply as similar check to the link above (but I am not sure we need one).
 
Please validate it with both default and service account authentication.

aozarov commented:
Looks good. One cleanup suggestion that needs to be validated.

ajkannan commented:
The compute credentials cleanup suggestion worked, so I added a commit for it.  I also noticed that service credentials didn't work with Datastore (I had been testing with storage earlier) because the v1beta2 dependency pulls in an outdated version of `com.google.api-client:google-api-client`.  I added a commit that fixes this as well.

Also, two cleanup questions:

1. There's `GoogleCredential` from `com.google.api-client:google-api-client` and there's `GoogleCredentials` (with an 's') `from google-auth-library-java`.  Could we use the `google-auth-library-java`'s GoogleCredentials [here](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-core/src/main/java/com/google/gcloud/AuthCredentials.java#L332)?  It looks like both have the `fromStream`.  Then we can get rid of the `com.google.api-client` dependency in the PR that fixes #403 (once the HttpInitializer code that uses `GoogleCredential` is gone).
2. We have a `ServiceAccountCredentials` class, but `google-auth-library-java` [does as well](https://github.com/google/google-auth-library-java/blob/master/oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java).  Do you think it'd be worthwhile to use `google-auth-library-java`'s instead?  It seems a bit more heavyweight, but it is a subclass of GoogleCredentials, which could be helpful for more auth restructuring later.

aozarov commented:
Yes, we should use `google-auth-library-java` `Credentials`instead of the `GoogleCredential` from `google-api-client`(however we do need that artifcat).

We can't use `ServiceAccountCredentials`from `google-auth-library-java` directly (as it is not restorable) however we can delegate to it.
